File: webapp/src/lib/about.ts
// Who made the tool, where it lives, and which build is serving this page.
//
// The identity constants MIRROR matcher/src/matcher/about.py, which is the
// single source of truth; matcher/tests/test_about.py fails if the two drift.
// For a results package, prefer the `provenance` that travels with the run
// (MatchOutput.provenance): it reports the ENGINE version that actually did
// the matching, which is what a reader of the report needs.

export const TOOL_NAME = "NeighborhoodMatcher";
/** Engine version this build ships (mirrors matcher/about.py VERSION). */
export const MATCHER_VERSION = "0.8.7";
export const AUTHORS = ["Dr. Benson Ku", "Dr. Patrick Kastner"] as const;
export const ORGANIZATION = "Sustainable Urban Systems Lab";
export const REPO_URL =
  "https://github.com/SustainableUrbanSystemsLab/NeighborhoodMatcher";
export const SITE_URL = "https://nbhdmatch.netlify.app/";
export const ORGANIZATION_URL = "https://sustainableurbansystems.com/";
/** Mirrors matcher/about.py AUTHOR_URLS — one entry per AUTHORS name that has a public profile. */
export const AUTHOR_URLS: Partial<Record<(typeof AUTHORS)[number], string>> = {
  "Dr. Benson Ku": "https://med.emory.edu/directory/profile/?u=BSKU",
  // Same URL as ORGANIZATION_URL, written literally so the object-literal
  // parity test (test_webapp_mirror_matches_python_author_urls) can see it —
  // it reads quoted-string pairs, not identifier references.
  "Dr. Patrick Kastner": "https://sustainableurbansystems.com/",
};

/** Authors as one display string: "A and B" / "A, B, and C". */
export const AUTHORS_LINE =
  AUTHORS.length === 2
    ? `${AUTHORS[0]} and ${AUTHORS[1]}`
    : AUTHORS.length < 2
      ? AUTHORS.join("")
      : `${AUTHORS.slice(0, -1).join(", ")}, and ${AUTHORS[AUTHORS.length - 1]}`;

/** The deployed build (see vite.config.ts `define`). */
export const BUILD = {
  version: __APP_VERSION__,
  commit: __APP_COMMIT__,
  builtAt: __BUILD_TIME__,
} as const;

/** e.g. "v0.1.0 (build 19b1b35)" — omits an unknown commit. */
export function buildLabel(): string {
  return BUILD.commit && BUILD.commit !== "unknown"
    ? `v${BUILD.version} (build ${BUILD.commit})`
    : `v${BUILD.version}`;
}

/** ISO-8601 UTC to the second: "2026-08-22T18:30:05Z". */
export function utcTimestamp(at: Date): string {
  return at.toISOString().replace(/\.\d+Z$/, "Z");
}

/**
 * The same instant in the viewer's local time, with its UTC offset —
 * "2026-08-22 14:30:05 UTC-04:00". This is the "when I ran this" a
 * researcher recognizes; the UTC stamp is the one that sorts.
 */
export function localTimestamp(at: Date): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  const date = `${at.getFullYear()}-${pad(at.getMonth() + 1)}-${pad(at.getDate())}`;
  const time = `${pad(at.getHours())}:${pad(at.getMinutes())}:${pad(at.getSeconds())}`;
  // getTimezoneOffset() is minutes BEHIND UTC, so the sign flips.
  const offsetMin = -at.getTimezoneOffset();
  const sign = offsetMin < 0 ? "-" : "+";
  const abs = Math.abs(offsetMin);
  return `${date} ${time} UTC${sign}${pad(Math.floor(abs / 60))}:${pad(abs % 60)}`;
}

/**
 * Local date/time as a filename-safe prefix: "20260904-1704". Minute
 * precision, no separators within each part — sorts chronologically and
 * survives every filesystem, unlike localTimestamp's colons and spaces.
 */
export function filenameTimestamp(at: Date): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  const date = `${at.getFullYear()}${pad(at.getMonth() + 1)}${pad(at.getDate())}`;
  const time = `${pad(at.getHours())}${pad(at.getMinutes())}`;
  return `${date}-${time}`;
}

File: webapp/public/matcher/about.py
"""
Identity and provenance of the tool: who wrote it, which version produced a
given set of results, and where to find the source.

Single source of truth. The webapp mirrors these values in
webapp/src/lib/about.ts (kept honest by tests/test_about.py) and gets the
authoritative copy at runtime from the `provenance` key that
web_api.coordinate_in_memory / assemble_results return — so a results
package always reports the version of the engine that actually ran, not
whatever the page was built with.
"""

from datetime import datetime, timezone

TOOL_NAME = "NeighborhoodMatcher"
VERSION = "0.8.4"
AUTHORS = ("Dr. Benson Ku", "Dr. Patrick Kastner")
ORGANIZATION = "Sustainable Urban Systems Lab"
REPO_URL = "https://github.com/SustainableUrbanSystemsLab/NeighborhoodMatcher"
SITE_URL = "https://nbhdmatch.netlify.app/"
ORGANIZATION_URL = "https://sustainableurbansystems.com/"
# One entry per AUTHORS name that has a public profile to link to; an author
# with no entry is credited by name only. Keyed by the exact AUTHORS string
# so a typo here fails loudly (test_about.py) rather than silently un-linking.
AUTHOR_URLS = {
    "Dr. Benson Ku": "https://med.emory.edu/directory/profile/?u=BSKU",
}

__version__ = VERSION


def authors_line():
    """Authors as one display string: 'A and B' / 'A, B, and C'."""
    names = list(AUTHORS)
    if len(names) <= 1:
        return "".join(names)
    if len(names) == 2:
        return f"{names[0]} and {names[1]}"
    return ", ".join(names[:-1]) + f", and {names[-1]}"


def utc_timestamp(moment=None):
    """
    Run timestamp as an ISO-8601 UTC string, seconds precision
    ('2026-08-22T18:30:05Z'). moment: an aware datetime, or None for now.
    """
    if moment is None:
        moment = datetime.now(timezone.utc)
    return moment.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def local_timestamp(moment=None):
    """
    The same instant in the machine's local time, with its UTC offset —
    what a researcher recognizes as 'when I ran this'. Falls back to the
    UTC string when the platform reports no local zone (some sandboxes).
    """
    if moment is None:
        moment = datetime.now(timezone.utc)
    local = moment.astimezone()
    stamp = local.strftime("%Y-%m-%d %H:%M:%S")
    offset = local.strftime("%z")
    if not offset:
        return utc_timestamp(moment)
    return f"{stamp} UTC{offset[:3]}:{offset[3:]}"


def provenance(moment=None):
    """
    Identity of the engine, ready to be written into a results package.

    Returns a JSON-safe dict; `generated_at_utc` / `generated_at_local`
    are omitted when moment is False (the browser stamps its own local
    time — Pyodide's clock has no timezone of its own).
    """
    info = {
        "tool": TOOL_NAME,
        "version": VERSION,
        "authors": list(AUTHORS),
        "authors_line": authors_line(),
        "organization": ORGANIZATION,
        "repo_url": REPO_URL,
        "site_url": SITE_URL,
    }
    if moment is not False:
        info["generated_at_utc"] = utc_timestamp(moment)
        info["generated_at_local"] = local_timestamp(moment)
    return info


def provenance_rows(moment=None, extra=None):
    """
    Provenance as ordered (key, value) rows for a CSV or a text header.
    extra: optional list of (key, value) run settings appended at the end.
    """
    info = provenance(moment)
    rows = [
        ("tool", info["tool"]),
        ("tool_version", info["version"]),
        ("authors", "; ".join(info["authors"])),
        ("organization", info["organization"]),
        ("repository", info["repo_url"]),
    ]
    if "generated_at_utc" in info:
        rows.append(("generated_at_utc", info["generated_at_utc"]))
        rows.append(("generated_at_local", info["generated_at_local"]))
    rows.extend(extra or [])
    return rows
